lxd/device/cdi/id.go

```diff
@@ -0,0 +1,126 @@
+package cdi
+
+import (
+	"fmt"
+	"strings"
+
+	"tags.cncf.io/container-device-interface/pkg/parser"
+)
+
+// Vendor represents the compatible CDI vendor.
+type Vendor string
+
+const (
+	// Nvidia represents the Nvidia CDI vendor.
+	Nvidia Vendor = "nvidia.com"
+)
+
+// ToVendor converts a string to a CDI vendor.
+func ToVendor(vendor string) (Vendor, error) {
+	switch vendor {
+	case string(Nvidia):
+		return Nvidia, nil
+	default:
+		return "", fmt.Errorf("Invalid CDI vendor (%q)", vendor)
+	}
+}
+
+// Class represents the compatible CDI class.
+type Class string
+
+const (
+	// GPU is a single discrete GPU.
+	GPU Class = "gpu"
+	// IGPU is an integrated GPU.
+	IGPU Class = "igpu"
+	// MIG is a single MIG compatible GPU.
+	MIG Class = "mig"
+)
+
+// ToClass converts a string to a CDI class.
+func ToClass(c string) (Class, error) {
+	switch c {
+	case string(GPU):
+		return GPU, nil
+	case string(IGPU):
+		return IGPU, nil
+	case string(MIG):
+		return MIG, nil
+	default:
+		return "", fmt.Errorf("Invalid CDI class (%q)", c)
+	}
+}
+
+// ID represents a Container Device Interface (CDI) identifier.
+//
+// +------------+-------+------------------------------+
+// |   Vendor   | Class |           Name               |
+// +---------------------------------------------------+
+// | nvidia.com |  gpu  | [dev_idx] or `all`           |
+// |            |  mig  | [dev_idx]:[mig_idx] or `all` |
+// |            |  igpu | [dev_idx] or `all`           |
+// +------------+-------+------------------------------+
+//
+// Examples:
+//   - nvidia.com/gpu=0
+//   - nvidia.com/gpu=all
+//   - nvidia.com/mig=0:1
+//   - nvidia.com/igpu=0
+type ID struct {
+	Vendor Vendor
+	Class  Class
+	Name   []string
```

Comment:
it shouldn't. Initially, I thought that this would be needed for the MIG case (index 0 for the device index, index 1 for the MIG index), but I'll let it as a string and extract the info later if a notation turns out to be MIG.